src/server/pkg/storage/fileset/index/reader.go

```diff
@@ -88,38 +148,123 @@ func (r *Reader) Next() (*Header, error) {
 		if strings.Compare(fullHdr.Idx.Range.LastPath, r.prefix) < 0 {
 			continue
 		}
-		// If a header with the prefix cannot show up after the current header,
-		// then we are done.
-		cmpSize := int64(math.Min(float64(len(hdr.Name)), float64(len(r.prefix))))
-		if strings.Compare(hdr.Name[:cmpSize], r.prefix[:cmpSize]) > 0 {
-			if err := r.Close(); err != nil {
+		return fullHdr, nil
+	}
+}
+
+// Peek peeks ahead in the index.
+func (r *Reader) Peek() (*Header, error) {
+	if err := r.setupLevels(); err != nil {
+		return nil, err
+	}
+	if r.levels[0].peekHdr != nil {
+		return r.levels[0].peekHdr, nil
+	}
+	var err error
+	r.levels[0].peekHdr, err = r.Next()
+	return r.levels[0].peekHdr, err
+}
+
+func (r *Reader) callback(level int) chunk.ReaderFunc {
+	return func() ([]*chunk.DataRef, error) {
+		hdr, err := r.next(level - 1)
+		if err != nil {
+			return nil, err
+		}
+		r.levels[level-1].currHdr = hdr
+		return hdr.Idx.DataOp.DataRefs, nil
+	}
+}
+
+// Copy is the basic data structure to represent a copy of data from
+// a reader to a writer.
+type Copy struct {
+	level int
+	raw   *chunk.Copy
+	hdrs  []*Header
+}
```

Comment:
Some thing with this copy struct as the other one, it seems really weird to have a public struct, that's returned by a public function but doesn't have any public fields. It also seems quite weird that there's two structs called `Copy` with identical comments describing them, but nothing in common between them field wise. I'm guessing that this one is for indexes while the other is for chunks, but I think the comments should clarify this more than they are.

Reply:
I don't really understand why that seems weird. It is basically a clean and well defined way to pass information through abstraction layers. Also, I think it would be good to view the copy structs with package scoping in mind. I will add some more documentation of the copy process and the relevant structs, but it does not make sense to me to specify that is an chunk/index copy struct when that should be apparent by the packaging.

Reply:
I didn't realize the package naming differences when reading, that makes sense then about the names being the same. I still had a few more questions about whether the struct was necessary on other other thread. But naming wise I'm satisfied so I'm going to resolve this one.